File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys


def get_parent(p, x):
    if p[x] != x:
        p[x] = get_parent(p, p[x])
    return p[x]


def union_parent(p, a, b):
    a = get_parent(p, a)
    b = get_parent(p, b)
    if a > b:
        p[a] = b
    elif a < b:
        p[b] = a
    else:
        return


def solution():
    v, e = map(int, sys.stdin.readline().rstrip().split())
    adj, parent, ans = [], [i for i in range(v + 1)], 0
    for i in range(e):
        a, b, c = map(int, sys.stdin.readline().rstrip().split())
        adj.append([a, b, c])
    adj.sort(key=lambda x: x[2])
    for v1, v2, dis in adj:
        if get_parent(parent, v1) != get_parent(parent, v2):
            union_parent(parent, v1, v2)
            ans += dis
    print(ans)


if __name__ == "__main__":
    solution()

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int v, e;
int parent[10001];
int get_parent(int parent[], int x) {
	if (parent[x] == x) return x;
	return parent[x] = get_parent(parent, parent[x]);
}
void union_parent(int parent[], int a, int b) {
	a = get_parent(parent, a);
	b = get_parent(parent, b);
	if (a < b)		parent[b] = a;
	else if (a > b) parent[a] = b;
	else			return;
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
	vector<vector<int>> adj;
	int ans = 0;
	cin >> v >> e;
	for (int i = 1; i < v + 1; i++)
		parent[i] = i;
	for (int i = 0; i < e; i++) {
		int v1, v2, dis;
		vector<int> tmp;
		cin >> v1 >> v2 >> dis;
		tmp.push_back(dis);
		tmp.push_back(v1);
		tmp.push_back(v2);
		adj.push_back(tmp);
	}
	sort(adj.begin(), adj.end());
	for (auto elem : adj) {
		if (get_parent(parent, elem[1]) != get_parent(parent, elem[2])) {
			union_parent(parent, elem[1], elem[2]);
			ans += elem[0];
		}
	}
	cout << ans;
	return 0;
}